rust/agents/validator/src/submit.rs

```diff
@@ -84,12 +84,12 @@ impl ValidatorSubmitter {
             };
 
             // ingest available messages from DB
-            while let Some(message) = self
+            while let Some(insertion) = self
                 .message_db
-                .retrieve_message_by_nonce(tree.count() as u32)?
+                .retrieve_merkle_tree_insertion_by_leaf_index(&(tree.count() as u32))?
             {
-                debug!(index = message.nonce, "Ingesting leaf to tree");
-                let message_id = message.id();
+                debug!(index = insertion.index(), queue_length = checkpoint_queue.len(), "Ingesting leaf to tree");
+                let message_id = insertion.message_id();
```

Comment:
can we explicitly track `index = tree.count()`

Reply:
Why not via the insertion index?